### PR TITLE
add eviction on cpu pressure

### PR DESCRIPTION
**Pull request by sjenning**
This adds eviction of "no CPU pods" once the shared CPU is depleted

cc @derekwaynecarr

### CODE REVIEW

**Comment by sjenning** [comment]
@ConnorDoyle ok, moved the eviction code to the manager

**Comment by sjenning** [comment]
@ConnorDoyle added test

File: pkg/kubelet/cpumanager/cpu_manager.go [hunk 1]
```diff
@@ -273,3 +304,61 @@ func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitR
...
+	return noCPUPods
+}
+
+func (m *manager) evictNoCPUPods() error {
+	noCPUPods := getNoCPUPods(m.getPodsFunc())
+	if len(noCPUPods) == 0 {
+		return nil
+	}
+
+	glog.Infof("[cpumanager] attempting to evict pods with no CPU request because shared CPU pool is empty: %v", noCPUPods)
```
Comment (ConnorDoyle) [inline pkg/kubelet/cpumanager/cpu_manager.go]: Nit: technically we are evicting because there is CPU pressure (although with static policy this only happens when shared pool is empty)

File: pkg/kubelet/cpumanager/cpu_manager.go [hunk 2]
```diff
@@ -273,3 +304,61 @@ func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitR
...
+			Phase:   v1.PodFailed,
+			Message: message,
+			Reason:  reason,
+		}
+		// record that we are evicting the pod
+		m.recorder.Eventf(pod, v1.EventTypeWarning, reason, message)
+		// this is a blocking call and should only return when the pod and its containers are killed.
+		err := m.killPodFunc(pod, status, nil)
+		if err != nil {
+			return fmt.Errorf("[cpumanager]: pod %s failed to evict %v", format.Pod(pod), err)
```
Comment (ConnorDoyle) [inline pkg/kubelet/cpumanager/cpu_manager.go]: Could we log and continue instead? Otherwise one container that the CRI can't evict for some reason could block the rest of the evictions.

File: pkg/kubelet/cpumanager/cpu_manager.go [hunk 3]
```diff
@@ -191,6 +214,14 @@ func (m *manager) reconcileState() (success []reconciledContainer, failure []rec
 	success = []reconciledContainer{}
 	failure = []reconciledContainer{}
 
+	if m.IsUnderCPUPressure() {
+		// Make sure all no-CPU pods are evicted.
+		// This really should only do something on the first reconcilation
+		// pass after transitioning to UnderPressure conditions, if all
+		// no-CPU pods are evicted successfully.
+		m.evictNoCPUPods()
```
Comment (ConnorDoyle) [inline pkg/kubelet/cpumanager/cpu_manager.go:222]: Suggest either to handle the error here instead of discarding, or remove the error from the signature.

**Review by ConnorDoyle — COMMENTED** [review]
Could we also add a test case for a burstable pod with one container that has CPU request and another container that has none?

Besides this and minor comments below LGTM

**Comment by sjenning** [comment]
@ConnorDoyle updated

File: pkg/kubelet/cpumanager/cpu_manager.go [hunk 4]
```diff
@@ -273,3 +304,61 @@ func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitR
...
+	for _, container := range pod.Spec.Containers {
+		if isNoCPUContainer(&container) {
+			return true
+		}
+	}
+	return false
+}
+
+func isNoCPUContainer(container *v1.Container) bool {
+	for resourceName, request := range container.Resources.Requests {
```
Comment (ConnorDoyle) [inline pkg/kubelet/cpumanager/cpu_manager.go]: Since container.Resources.Requests is a map, want to do a lookup here instead of iterating the entries?

**Review by ConnorDoyle — APPROVED** [review]

File: pkg/kubelet/cpumanager/cpu_manager.go (hunk 4 above)
Comment (sjenning) [reply pkg/kubelet/cpumanager/cpu_manager.go]: ah yes, that will make this simpler :+1:

**Comment by sjenning** [comment]
@ConnorDoyle updated and added the 2 container test pod I forgot before.

**Comment by ConnorDoyle** [comment]
💯
